Provide non-headless (xvfb) UI tests in CI using ide-probe

PawelLipski opened the pull request:
~Almost ready, but can't be merged in the current state since the driver fat jar (ca. 25MB) is committed directly in the repository... which will forever pollute the master history, even if the jar is removed on a later commit.~ => ok, resolved

**Note:** some non-essential follow-up tasks are already listed in https://github.com/VirtusLab/git-machete-intellij-plugin/issues/274

PawelLipski commented on uiTests/build.gradle:
```diff
@@ -0,0 +1,22 @@
+}
+
+dependencies {
+  testImplementation fileTree(dir: 'lib', include: '*.jar')
```
**TODO before merge:** don't keep the driver jar in the repo

PawelLipski commented on README.md:
```diff
@@ -50,16 +54,32 @@ To enable logging in `DEBUG` level, add selected categories to list in `Help` ->
+
+### Run UI tests
+
+This repo ships with a JUnit driver fat jar (see `uiTests/lib/junit-driver-assembly-0.1.jar`) for UI tests,
```
Noooopes get rid of this :bangbang:

PawelLipski commented on README.md:
````diff
@@ -50,16 +54,32 @@ To enable logging in `DEBUG` level, add selected categories to list in `Help` ->
+UI tests are by default excluded from `./gradlew test` unless `runUiTests` property is set:
+
+```
+./gradlew -PrunUiTests test
````
@mkondratek @micpiotrowski if you could run this stuff on your own machines

mkondratek replied:
see my comment below

PawelLipski replied:
Comments addressed, let's see if @micpiotrowski will have any issues if he follows the updated guidelines

micpiotrowski replied:
`./gradlew -PrunUiTests test` ran without any issue

micpiotrowski replied:
but it was relatively fast (about one minute with compilation) so I'm not sure if everything went ok (was tested)

PawelLipski replied:
This file is outdated... pls take a look into the updated README, it's `./gradlew -PenableUiTests --info test` now

PawelLipski replied:
And it should open IntelliJ twice (once for 2019.3, and once for 2020.1) during the test run

micpiotrowski replied:
Yeah, my fault.
When I done this like in README everything goes fine

PawelLipski commented on .circleci/config.yml:
```diff
@@ -42,15 +42,39 @@ jobs:
+      - run:
+          name: Run UI tests
+          command: |
+            ### Remove once Docker image is regenerated
```
Anybody wanna do a good deed? ;)

micpiotrowski replied:
In progress...

PawelLipski replied:
Ok pls ping me when ready

micpiotrowski replied:
Ready

PawelLipski commented on .dockerignore:
```diff
@@ -12,6 +13,7 @@ Dockerfile
 .gitignore
 .gradle/
 .idea/
+**/lib/
```
To be removed once the fat jar is removed from the repo

PawelLipski commented on uiTests/src/test/scala/com/virtuslab/gitmachete/uitest/OpenTabActionUiTest.scala:
```diff
@@ -0,0 +1,42 @@
+import com.virtuslab.ideprobe.{ IntegrationTestSuite, IntelliJFixture, WorkspaceConfig, WorkspaceTemplate }
+import org.junit.{ Assert, Test }
+
+class OpenTabActionUiTest extends IntegrationTestSuite {
```
Just too much hassle writing that in Java (using code compiled in Java from Scala is trivial; the reverse is possible but generally PITA)

PawelLipski commented on frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/ideprobe/OpenTabAction.java:
```diff
@@ -0,0 +1,44 @@
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
+
+/**
+ * This action is only used for UI tests.
```
Or maybe... it can be just left as is in production builds as well? :thinking:

mkondratek commented:
_My test execution summary_

Issues:
- executing gradle task from command line (I had java 8)
- repo building script (needs manual run)
- executing tests (wrong path to repo)
- fetching intellij takes time (without -info, I thought it hanged) 

Tests :
- works :tada: 

Additional observations:
- omitting projects windows (the one where you chose the project :xd:) could speed up the tests
- the same about tip popup within IDE

PawelLipski commented:
> executing gradle task from command line (I had java 8)
Added a note to README.md

> repo building script (needs manual run)
Fixed

> executing tests (wrong path to repo)
Fixed

> fetching intellij takes time (without -info, I thought it hanged)
Added a note to README.md

mkondratek commented on src/main/resources/META-INF/plugin.xml:
```diff
@@ -99,5 +99,10 @@
         </group>
+
+        <!-- TODO (#274): consider excluding this action from the production builds of the plugin -->
+        <action id="GitMachete.OpenTabAction"
```
This action might actually be useful. Idea log as an example:

![image](https://user-images.githubusercontent.com/19799111/82117021-e8bab100-976d-11ea-9989-4876ab8eed53.png)
 
Commented in https://github.com/VirtusLab/git-machete-intellij-plugin/issues/274#issuecomment-629621134 too.

micpiotrowski commented on .circleci/config.yml:
```diff
@@ -42,15 +42,60 @@ jobs:
+            echo "deb https://dl.bintray.com/sbt/debian /" | tee -a /etc/apt/sources.list.d/sbt.list
+            curl -sL "https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x2EE0EA64E40A89B84B2DF73499E82A75642AC823" | apt-key add
+            apt-get update
+            apt-get install -y sbt
```
Can't things up to this line be baked into Docker image?
It looks like regular software installation from repo.

PawelLipski replied:
Yeah, could be... but I (naively) hope we soon won't need them anymore (i.e. we won't need to build ide-probe within the CI pipeline at all, but we'll use externally published artifacts instead, and hence sbt won't be needed at all)
